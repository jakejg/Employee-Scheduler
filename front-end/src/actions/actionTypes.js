export const LOAD_JOBS = "LOAD_JOBS";
export const ADD_JOBS = "ADD_JOB";