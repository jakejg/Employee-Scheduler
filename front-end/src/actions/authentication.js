import axios from 'axios';
import { 
    ADD_OR_REMOVE_TOKEN
 } from './actionTypes';
import {addCompanyToAPI} from '../helpers/JobApi'
import {BASE_URL} from '../config';


export const register = (data) => {
    
    return async (dispatch) => {
        try {
            // designate admin user
            data.is_admin = true
            // create admin user
       
            const res = await axios.post(`${BASE_URL}/auth/register`, data);
       
            dispatch(addOrRemoveToken(res.data.token));
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const login = (data) => {
    return async (dispatch) => {
        try {
            let res = await axios.post(`${BASE_URL}/auth/login`, data);
            dispatch(addOrRemoveToken(res.data.token));
        }
        catch(e) {
            // console.log(e.response.data.message)
            return e.response.data.message
        }
    }
}

export const addOrRemoveToken = (token) => {
    return {type: ADD_OR_REMOVE_TOKEN, token}
}
