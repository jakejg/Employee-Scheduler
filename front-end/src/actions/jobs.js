import { 
    LOAD_JOBS,
    EDIT_JOB, 
    ADD_JOB,
    EDIT_JOB_STAFF,
    DELETE_JOB
 } from './actionTypes';
import {JobAPI} from '../helpers/JobApi';

export const loadJobsFromAPI = (comp_id) => {
    return async (dispatch) => {
        try {
            const jobData = await JobAPI.loadJobs(comp_id)
            dispatch(loadJobs(jobData));
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const loadJobs = (jobs) => {
    return {type: LOAD_JOBS, jobs}
}

export const addJobOnAPI = (jobToAdd) => {

    return async (dispatch) => {
        try {
            let job = await JobAPI.addJob(jobToAdd)
            dispatch(addJob(job.id, job))
            return {message: "Job created Successfully", severity: "success"}
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const addJob = (id, job) => {
    return {type: ADD_JOB, id, job}
}


export const getJobFromAPI = (ID) => {
    return async (dispatch) => {
        try {
            let job = await JobAPI.getJob(ID)
            dispatch(editJob(job.id, job))
        }
        catch(e) {
            console.log(e)
            return e.message
        }
    }
}

export const editJobOnAPI = (id, changedProps) => {
    return async (dispatch) => {
        try {
            let job = await JobAPI.editJob(id, changedProps)
            console.log(job)
            dispatch(editJob(job.id, job))
        }
        catch(e) {
            console.log(e)
            return e.message
        }
    }
}

export const editJob = (id, job) => {
    return {type: EDIT_JOB, id, job}
}

export const addStaffToJobOnAPI = (jobId, staffId) => {
    return async (dispatch) => {
        try {
            let data = await JobAPI.addStaffToJob(jobId, staffId)
            console.log(data)
            dispatch(editJobStaff(jobId, data.staffList, data.staff_filled))
            
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const removeStaffFromJobOnAPI = (jobId, staffId) => {
    return async (dispatch) => {
        try {
            let data = await JobAPI.removeStaffFromJob(jobId, staffId)
            dispatch(editJobStaff(jobId, data.staffList, data.staff_filled))
            
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const editJobStaff = (id, staff, staff_filled) => {
    return {type: EDIT_JOB_STAFF, id, staff, staff_filled}
}

export const deleteJobOnAPI = (id) => {
    return async (dispatch) => {
        try {
            let msg = await JobAPI.deleteJob(id)
            dispatch(deleteJob(id))
        }
        catch(e) {
            console.log(e)
            return e.message
        }
    }
}

export const deleteJob = (id) => {
    return {type: DELETE_JOB, id}
}