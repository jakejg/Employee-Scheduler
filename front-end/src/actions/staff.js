import { 
    LOAD_STAFF,
    EDIT_STAFF,
    ADD_STAFF
 } from './actionTypes';
import { StaffAPI } from '../helpers/StaffApi';

export const loadStaffFromAPI = (comp_id) => {
    return async (dispatch) => {
        try {
            let staffData = await StaffAPI.loadStaff(comp_id);
            dispatch(loadStaff(staffData));
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const loadStaff = (staff) => {
    return {type: LOAD_STAFF, staff}
}


export const getStaffFromAPI = (ID) => {
    return async (dispatch) => {
        try {
            let staff = await StaffAPI.getStaff(ID)
        
            dispatch(editStaff(staff.id, staff))
            }
        catch(e) {
            console.log(e)
        }
    }
}

export const editStaff = (id, staff) => {
    return {type: EDIT_STAFF, id, staff}
}

export const addStaffOnAPI = (staffToAdd) => {
   
    return async (dispatch) => {
        try {
            const staff = await StaffAPI.addStaff(staffToAdd) ;
            dispatch(addStaff(staff.id, staff))
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const addStaff = (id, staff) => {
    return {type: ADD_STAFF, id, staff}
}