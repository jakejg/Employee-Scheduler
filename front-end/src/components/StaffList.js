import React from 'react';
import { useSelector } from 'react-redux';
import {List, ListItem, ListItemText, Box, Typography, makeStyles, Chip } from '@material-ui/core';
import Loading from './Loading';

const useStyles = makeStyles(() => ({
    text: {
        display: 'flex',
        justifyContent: 'start',
        width: '125px'
    },
    box: {
        display: 'flex',
        justifyContent: 'center',
    }
}))


const StaffList = () => {
    const staff = useSelector(state => state.staff);
    const classes = useStyles();
    const loading = !Object.keys(staff).length
    
    if (loading) return <Loading />

    const working = [];
    const available = [];

    for (let id in staff){
        if (staff[id].isWorking) {
            working.push({name: `${staff[id].first_name} ${staff[id].last_name}`, id})
        }
        else available.push({name: `${staff[id].first_name} ${staff[id].last_name}`, id});
    }
    return (
        <div>
            <Typography variant='h4' align='center'>Staff Currently Working</Typography>
                <List>
                    {working.map(staff => <ListItem key={staff.id} className={classes.box}>
                                            <Box  className={classes.text}>
                                                <Chip 
                                                    label={staff.name} 
                                                    component="a" href={`staff/${staff.id}`}
                                                    color="secondary"
                                                    clickable 
                                                    />
                                            </Box>
                                        </ListItem>)}
                </List>
            <Typography variant='h4' align='center'>Available Staff</Typography>
                <List>
                    {available.map(staff => <ListItem key={staff.id} className={classes.box}>
                                                <Box  className={classes.text}>
                                                    <Chip 
                                                        label={staff.name} 
                                                        component="a" href={`staff/${staff.id}`}
                                                        color="secondary"
                                                        clickable 
                                                    />   
                                                </Box>
                                            </ListItem>)}
                </List>

        </div>
    );
}

export default StaffList;
