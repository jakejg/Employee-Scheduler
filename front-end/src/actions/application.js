import {CHANGE_DRAWER, CHANGE_NO_JOBS, CHANGE_NO_STAFF} from './actionTypes';

export const changeDrawer = () => {
    return {type: CHANGE_DRAWER}
}

export const changeNoJobs = () => {
    return {type: CHANGE_NO_JOBS}
}

export const changeNoStaff = () => {
    return {type: CHANGE_NO_STAFF}
}