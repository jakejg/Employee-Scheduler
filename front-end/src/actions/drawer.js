import {CHANGE_DRAWER} from './actionTypes';

export const changeDrawer = () => {
    return {type: CHANGE_DRAWER}
}