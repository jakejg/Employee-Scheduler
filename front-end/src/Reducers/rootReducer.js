import jobs from './jobsReducer';
import staff from './staffReducer';
import { combineReducers } from "redux";

export default combineReducers({jobs, staff})