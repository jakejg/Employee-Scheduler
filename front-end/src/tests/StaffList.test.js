import React from 'react';
import { render } from '@testing-library/react';
import {Provider} from 'react-redux'
import StaffList from '../components/StaffList';
import {createStore, applyMiddleware} from "redux";
import rootReducer from "../reducers/rootReducer";
import { MemoryRouter } from "react-router";
import { loadStaff } from '../actions/staff';
import thunk from 'redux-thunk'

const staff = {
    1 : {
        first_name: "Stacey",
        last_name: "Lopez"
    },
    2 : {
        first_name: "Jon",
        last_name: "Martin"
    }
}

const store = createStore( rootReducer, applyMiddleware(thunk));
store.dispatch(loadStaff(staff)); 

it('renders without crashing', () => {

  render(
    <MemoryRouter>
      <Provider store={store}>
      <StaffList />
      </Provider>
    </MemoryRouter>
)
});

it('matches snapshot', () => {
    const { asFragment } = render(
        <MemoryRouter>
            <Provider store={store}>
                <StaffList />
            </Provider>
        </MemoryRouter>
    )
    expect(asFragment()).toMatchSnapshot();
})


it('renders a list of staff', () => {
    const { getByText } = render(
        <MemoryRouter>
            <Provider store={store}>
                <StaffList />
            </Provider>
        </MemoryRouter>
    )
    expect(getByText('Stacey Lopez')).toBeInTheDocument();
    expect(getByText('Jon Martin')).toBeInTheDocument();
});