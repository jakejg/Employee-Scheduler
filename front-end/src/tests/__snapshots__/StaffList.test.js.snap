// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h4
      class="MuiTypography-root MuiTypography-h4 MuiTypography-alignCenter"
    >
      Staff Currently Working
    </h4>
    <ul
      class="MuiList-root MuiList-padding"
    />
    <h4
      class="MuiTypography-root MuiTypography-h4 MuiTypography-alignCenter"
    >
      Available Staff
    </h4>
    <ul
      class="MuiList-root MuiList-padding"
    >
      <li
        class="MuiListItem-root makeStyles-box-6 MuiListItem-gutters"
      >
        <div
          class="MuiBox-root MuiBox-root-7 makeStyles-text-5"
        >
          <a
            class="MuiChip-root MuiChip-colorSecondary MuiChip-clickableColorSecondary MuiChip-clickable"
            href="staff/1"
            role="button"
            tabindex="0"
          >
            <span
              class="MuiChip-label"
            >
              Stacey Lopez
            </span>
          </a>
        </div>
      </li>
      <li
        class="MuiListItem-root makeStyles-box-6 MuiListItem-gutters"
      >
        <div
          class="MuiBox-root MuiBox-root-8 makeStyles-text-5"
        >
          <a
            class="MuiChip-root MuiChip-colorSecondary MuiChip-clickableColorSecondary MuiChip-clickable"
            href="staff/2"
            role="button"
            tabindex="0"
          >
            <span
              class="MuiChip-label"
            >
              Jon Martin
            </span>
          </a>
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
